feat(text-input): adds fluid variation of text input

aledavila opened the pull request:
References #5416

Adds fluid form component
Fluid form will handle fluid for all of the components in form

#### Changelog

**New**

- fluid text input variation
- add new `FluidForm` component


Check text inputs work and look as intended. Check fluid text input matches.

Commit:
feat(text-inputs): adds fluid variation of text input

Commit:
feat(text-input): add fluid to props

joshblack commented on packages/react/src/components/TextInput/TextInput.js:
```diff
@@ -185,6 +193,11 @@ TextInput.propTypes = {
+  /**
+   * `true` to use fluid variation of input field.
+   */
+  fluidInput: PropTypes.bool,
```
Do we know if forms with fluid inputs are supposed to be constructed? Is it an all-or-nothing type thing? (meaning all fluid versus not) or is it more mix-and-match?

aledavila replied:
I think its all or nothing according this the issue at the very bottom is the example of the fluid form https://github.com/carbon-design-system/carbon/issues/5416

joshblack replied:
Would it be helpful to set this at the form level then if it's all-or-nothing? Could help to avoid repeating `fluidInput` repeatedly and accidentally mixing-and-matching. For example:

```jsx
<Form fluid>
  <TextInput />
</Form>

// Or
<FluidForm>
  <TextInput />
</FluidForm>
```

tw15egan replied:
I believe the fluid inputs can be used outside the context of a Form. We could always add Form level controls as an enhancement in the future, so you don't have to toggle it on for each form item.

joshblack replied:
Ah got it, that makes sense. Is fluid basically like another size that folks can choose from basically?

tw15egan replied:
Nevermind! According to Design it should only be used in the context of a Form or Title. Since that is the case, it would make more sense to add a prop to the Form and Tile components to handle toggling this feature on / off

tw15egan commented on packages/components/src/components/text-input/_text-input.scss:
```diff
@@ -64,6 +64,28 @@
+
+  .#{$prefix}--text-input-wrapper--fluid .#{$prefix}--text-input {
+    min-height: rem(64px);
+    padding: rem(32px) rem(14px) $carbon--spacing-05;
```
I believe we want `16px` of padding on the sides

```suggestion
    padding: rem(32px) $carbon--spacing-05 rem(13px);
```

tw15egan commented on packages/components/src/components/text-input/_text-input.scss:
```diff
@@ -64,6 +64,28 @@
+    position: absolute;
+    top: rem(14px);
+    left: rem(14px);
+    z-index: 1;
```
We can remove the `margin` here since it isn't doing anything 

```css
margin: 0;
```

tw15egan commented on packages/components/src/components/text-input/_text-input.scss:
```diff
@@ -64,6 +64,28 @@
+  .#{$prefix}--text-input-wrapper--fluid .#{$prefix}--label {
+    position: absolute;
+    top: rem(14px);
+    left: rem(14px);
```
```suggestion
    top: rem(13px);
    left: $carbon--spacing-05;
```

aledavila replied:
the reason I made this 14 instead of 16 is because for some reason if I put 16 it goes past aligning with the placeholder.

tw15egan replied:
It shouldn't if you adjust the text input padding to 16 as well, see https://github.com/carbon-design-system/carbon/pull/5830/files#r406366436

<img width="1005" alt="Screen Shot 2020-04-09 at 11 55 26 AM" src="https://user-images.githubusercontent.com/11928039/78930518-1dde3000-7a59-11ea-9f0a-1b22d0c6e807.png">

tw15egan commented on packages/react/src/components/TextInput/TextInput.js:
```diff
@@ -30,6 +30,7 @@ const TextInput = React.forwardRef(function TextInput(
     helperText,
     light,
     size,
+    fluidInput,
```
Perhaps just `fluid`?

tw15egan committed:
Merge branch 'master' into fluid-inputs

Commit:
Merge branch 'master' of github.com:carbon-design-system/carbon into fluid-inputs

Commit:
Merge branch 'fluid-inputs' of github.com:aledavila/carbon into fluid-inputs

Commit:
feat(form): adds fluid prop to form component

Commit:
chore(text-input): move fluid handling to form

tw15egan committed:
Merge branch 'master' into fluid-inputs

tw15egan committed:
feat(FluidForm): add fluid form component

aledavila committed:
Merge pull request #1 from tw15egan/fluid-form-addition

feat(FluidForm): add fluid form component

Commit:
Merge branch 'master' of github.com:carbon-design-system/carbon into fluid-inputs

Commit:
feat(text-input): add fluid error state

Commit:
chore(text-input): update invalid state css

tw15egan committed:
feat(fluid-input): add context to fluid form

aledavila committed:
Merge pull request #2 from tw15egan/fluid-form-context

feat(fluid-input): add context to fluid form

tw15egan committed:
Merge branch 'master' into fluid-inputs

aagonzales committed:
Merge branch 'master' into fluid-inputs

aagonzales reviewed (commented):
The text input looks good! In the fluid form the buttons should behave like the modal buttons and span the full width or half of the container.

![image](https://user-images.githubusercontent.com/11670886/79797848-a0020a80-831d-11ea-904c-d876ce99086e.png)


In the form example all the fields should be stack directly on top of each other. Can we remove the padding before the invalid field?
![image](https://user-images.githubusercontent.com/11670886/79797963-d50e5d00-831d-11ea-932b-23aa0d84ecba.png)

aledavila committed:
Merge branch 'master' into fluid-inputs

asudoh commented on packages/react/src/components/TextInput/TextInput.js:
```diff
@@ -83,8 +91,10 @@ const TextInput = React.forwardRef(function TextInput(
     <div className={helperTextClasses}>{helperText}</div>
   ) : null;
 
+  const { isFluid } = useContext(FormContext);
```
Wondering if we can we make this a prop... I imagine usage of context may stem from design guide dictating that fluid is per-form basis. OTOH software designs tend to go further, which is, setting API in different levels (layers). So we can break down "per-form fluid mode" for component implementation, one is a pure mean to change fluid/regular modes (e.g. `<TextInput layout="fluid">`), another is app-level mean to ensure all form items is the desired mode, e.g.:

```javascript
const AppTextInput = props => {
  const { useFluid } = useContext(FormContext);
  return <TextInput {...props} layout={useFluid ? 'fluid' : 'regular'} />;
};
```

tw15egan replied:
@asudoh we decided to use context because we wanted to avoid having the user set a `fluid` prop for each individual form element. Since they can only be used in the context of a `FluidForm`, we wanted to make it as easy as possible for users to create a fluid form, while not allowing much deviation from design.

asudoh replied:
I see, thank you for the background @tw15egan. I tend to avoid making "syntactic sugar" thing affect system design (UX-level requirement and system design are often different animals), but I at least now see where it came from.

Another approach to achieve such syntactic sugar is introducing `<FluidTextInput>`.

Commit:
Merge branch 'master' of github.com:carbon-design-system/carbon into fluid-inputs

Commit:
Merge branch 'fluid-inputs' of github.com:aledavila/carbon into fluid-inputs

Commit:
chore(fluid-form): remove button from storybook and margin from fluid

aledavila commented:
@aagonzales hey I removed the margin but also removed the button for now. We are slowly building the fluid form since more things need to be added in, which will happen in a separate PR. For now this mainly for the text input.

aagonzales reviewed (approved):
The preview link isn't working but since this PR is just about text input I'll go ahead and 👍

tw15egan committed:
Merge branch 'master' into fluid-inputs

joshblack commented on packages/react/src/components/FluidForm/FluidForm.js:
```diff
@@ -0,0 +1,46 @@
+  /**
+   * `true` to use fluid variation of input field.
+   */
+  fluid: PropTypes.bool,
```
Are we using this prop for anything? 👀

aledavila replied:
good catch. no we don't expose the fluid prop anymore.

joshblack commented on packages/components/src/components/text-input/_text-input.scss:
```diff
@@ -170,6 +170,60 @@
+  }
+
+  .#{$prefix}--form--fluid .#{$prefix}--text-input__invalid-icon {
+    top: 80px;
```
Would this need to be in `rem` potentially for scaling or is fixed better here? (Wasn't sure 👀)

```suggestion
    top: rem(80px);
```

aledavila replied:
I think so. The rest is in rem. Do you think it would have any positioning issues? Since I was going off the height of the input + the padding of the error message.

Commit:
Merge branch 'master' of github.com:carbon-design-system/carbon into fluid-inputs

joshblack commented on packages/react/src/components/FluidForm/FluidForm.js:
```diff
@@ -0,0 +1,46 @@
+
+const { prefix } = settings;
+
+function FluidForm({ className, children, ...other }) {
```
Do we think people may want to attach a `ref` to a form?? Wasn't sure if we do it for `Form` or not 🤔

aledavila replied:
hmm not sure. Its not done in `Form` though

Commit:
chore(text-input): css fixes

tw15egan committed:
Merge branch 'master' into fluid-inputs

joshblack commented on packages/react/src/components/TextInput/TextInput.js:
```diff
@@ -94,8 +104,10 @@ const TextInput = React.forwardRef(function TextInput(
           <WarningFilled16 className={`${prefix}--text-input__invalid-icon`} />
         )}
         {input}
+        <hr className={`${prefix}--text-input__divider`} />
```
Does the divider need to be conditional now based on `isFluid` or is this also an addition to the base text input as well?

aledavila replied:
No it has to be conditional. Only in isFluid.

Commit:
Merge branch 'master' of github.com:carbon-design-system/carbon into fluid-inputs

Commit:
Merge branch 'fluid-inputs' of github.com:aledavila/carbon into fluid-inputs

Commit:
chore(text-input): make divider conditional

dakahn reviewed (approved):
Looks good. Thanks and sorry for getting to this so late 🏄

tw15egan committed:
Merge branch 'master' into fluid-inputs

Commit:
chore(text-input): update snapshots

joshblack committed:
Merge branch 'master' into fluid-inputs

Commit:
Merge branch 'master' into fluid-inputs

Commit:
Merge branch 'master' into fluid-inputs

Commit:
Merge branch 'master' into fluid-inputs
